### DISCUSSION TITLE
How to access motion_library from utils folder using webot_ros2

### DISCUSSION THREAD

**Discussion by rua0ra1** [cyberbotics/wrestling#49] [Q&A] [answered]
Hi,
I am trying implement python bob example in ros2 using webots_ros2 package. I am trying to access motion_library.py from utils folder. Using the same sys.path function mention in bob python example. But, I am getting 
" ModuleNotFoundError: No module named 'utils' " error. I am wondering whether I am missing something here. I really appreciate it if you can help me with this.

following is the my folder structure.
├── motions
├── opponent
├── participant ( ros2 package)
│   ├── launch
│   ├── participant ( this folder has my controller.py file)
│   │   └── __pycache__
│   ├── resource
│   └── test
├── referee
└── utils
    └── __pycache__

**Comment by Jean-Eudes-le-retour** [comment, marked as answer]
The [utils](https://github.com/cyberbotics/wrestling-bob/tree/main/controllers/utils) folder inside the Bob controller is not an official Python module. It is simply provided to give people ideas for their controllers. If you want to reuse its code for your ROS 2 controller, you'll need to copy the `utils` folder to your package and to set your Python path correctly so that Python detects it.

**Reply by Jean-Eudes-le-retour** [reply]
You might be able to simply place `utils` and `motions` like this
```
├── opponent
├── participant ( ros2 package)
│ ├── launch
│ ├── motions
│ ├── utils
│ ├── participant
```
You should then be able to use the same Python code inside controller.py:
``` Python
import sys
sys.path.append('..')
from utils.motion_library import MotionLibrary
```

**Reply by rua0ra1** [reply]
Hi @Jean-Eudes-le-retour , Thanks for the answer. I have tried it but it didn't work :(.

Error message:-
   File "/home/ram/my_projects/cyberbots_ws/install/participant/lib/python3.10/site-packages/participant/nao_controller.py", line 4, in <module>
[driver-1]     from utils.motion_library import MotionLibrary
[driver-1] ModuleNotFoundError: No module named 'utils'

Directory structure:-
participant (ros package)
├── launch
├── motions
├── participant
│   └── __pycache__
├── resource
├── test
└── utils
    └── __pycache__

**Reply by Jean-Eudes-le-retour** [reply]
Aha, I found it, there are two things missing:

First thing, set up your `controllers` folder like this (notice that the motions folder is outside the participant package):

```
controllers
├── motions
├── participant (ros package)
│   ├── launch
│   ├── participant
│   ├── resource
│   ├── test
│   └── utils
└── wrestling_referee
```

### 1. Additional Python modules in a ROS 2 package

To add an additional Python module, we have to modify the [setup.py file](https://github.com/cyberbotics/wrestling-alice-ros-2/blob/main/controllers/participant/setup.py#L8) at the root of the ROS 2 package ([source](https://answers.ros.org/question/367793/including-a-python-module-in-a-ros2-package/))

You have to modify `packages=[package_name],` to `packages=[package_name, 'utils'],` to include the `utils` module in the site-package Python folder used by the ROS 2 package.

### 2. Weird webots_ros2_driver `Motion` workaround

Currently, `webots_ros2_driver` only provides an easy access to the `Robot()` Webots API with the [webots_node variable](https://github.com/cyberbotics/wrestling-alice-ros-2/blob/main/controllers/participant/participant/nao_controller.py#L6). But if we want to play motion files we need to access Webots' `Motion` class.

To have access to all Webots Python classes, we have to add at the top of the [nao_controller.py](https://github.com/cyberbotics/wrestling-alice-ros-2/blob/main/controllers/participant/participant/nao_controller.py):

``` Python
import os
from ament_index_python.packages import get_package_prefix
os.environ['WEBOTS_HOME'] = get_package_prefix('webots_ros2_driver')
```

With thoses two fixes, we can finally import the MotionLibrary with a standard import:

``` Python
from utils.motion_library import MotionLibrary
```

### Example nao_controller.py

If you have set up correctly the `motions` folder and the `setup.py` file, here is a `nao_controller.py` file that should work:

``` Python
import rclpy
import os
from ament_index_python.packages import get_package_prefix
os.environ['WEBOTS_HOME'] = get_package_prefix('webots_ros2_driver')
from utils.motion_library import MotionLibrary

class NaoDriver:
    def init(self, webots_node, properties):
        self.__library = MotionLibrary()

        rclpy.init(args=None)
        self.__node = rclpy.create_node('nao_driver')

    def step(self):
        # Mandatory function to go to the next simulation step
        rclpy.spin_once(self.__node, timeout_sec=0)

        self.__library.play('HandWave')
```

**Reply by Jean-Eudes-le-retour** [reply]
I'll update https://github.com/cyberbotics/wrestling-alice-ros-2 so that the controller behaves exactly like the other Alice example

**Reply by Jean-Eudes-le-retour** [reply]
I've made a new ROS 2 example which implements what I was talking about: https://github.com/cyberbotics/wrestling-bob-ros-2

**Comment by rua0ra1** [comment]
Hi @Jean-Eudes-le-retour , Thanks for detail answer and providing bob example in ros2. I have tried to run the given example in my local system. I am getting the following issue. I have tried some tweaks to resolve the issue .But, unfortunately, I could not resolve it
 error message:
 Traceback (most recent call last):
[driver-1]   File "/home/ram/projects/ros2_webots_practice_ws/install/participant/lib/python3.10/site-packages/participant/nao_controller.py", line 35, in step
[driver-1]     rclpy.spin_once(self.__node, timeout_sec=0)
[driver-1] AttributeError: 'NaoDriver' object has no attribute '_NaoDriver__node'. Did you mean: '_NaoDriver__robot'?

**Reply by Jean-Eudes-le-retour** [reply]
I've come across this error message by the past, and it was always due to another error happening before it, can you confirm that no other style of error message happened before this one?

I've tested the bob ros 2 example with multiple releases of Webots and had no issue.

**Reply by rua0ra1** [reply]
You are right @Jean-Eudes-le-retour , here is the actual problem
[driver-1] Traceback (most recent call last):
[driver-1]   File "/home/ram/projects/webotsros2bob_ws/install/participant/lib/python3.10/site-packages/participant/nao_controller.py", line 15, in init
[driver-1]     self.__library = MotionLibrary()
[driver-1]   File "/home/ram/projects/webotsros2bob_ws/install/participant/lib/python3.10/site-packages/utils/motion_library.py", line 25, in __init__
[driver-1]     for motion_file in os.listdir(motion_dir):
[driver-1] FileNotFoundError: [Errno 2] No such file or directory: '../motions/'

and also
I am definging following env variable in my bash file
export WEBOTS_HOME=/usr/local/webots

do you think it might be cauing the problem?

**Reply by Jean-Eudes-le-retour** [reply]
No, the problem doesn't come from the environment variable. The issue seems to be that the `motions` folder cannot be found (using the relative path `../motions/`). To make it work locally to your computer, you could copy the `motions` folder to `/home/ram/projects/` or you could modify the path in [utils/motion_library.py](https://github.com/cyberbotics/wrestling-bob-ros-2/blob/main/controllers/participant/utils/motion_library.py#L23) with an absolute path to your motions path.

Note that when you will submit your controller, you will need to change back the path to a relative path to make it work in the GitHub runner.

**Reply by rua0ra1** [reply]
Thanks @Jean-Eudes-le-retour it solved the issue and really appreciate it.
